semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/ClasspathIndex.scala

```diff
@@ -0,0 +1,138 @@
+package scala.meta.internal.metacp
+
+import java.nio.file.FileVisitResult
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.SimpleFileVisitor
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.jar.JarFile
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+import scala.meta.io.AbsolutePath
+import scala.meta.io.Classpath
+import scala.meta.internal.io.PathIO
+
+/** An index to lookup class directories and classfiles by their JVM names. */
+final class ClasspathIndex private (val dirs: collection.Map[String, Classdir]) {
+
+  override def toString: String = {
+    s"ClasspathIndex(${dirs.size} entries)"
+  }
+
+  /** Returns a classfile with the given path. */
+  def getClassfile(path: String): Option[Classfile] = {
+    getClassfile(PathIO.dirname(path), PathIO.basename(path))
+  }
+
+  /** Returns a classfile with the given directory and filename. */
+  def getClassfile(directory: String, filename: String): Option[Classfile] = {
+    dirs.get(directory) match {
+      case Some(pkg) =>
+        pkg.members.get(filename).collect {
+          case e: Classfile => e
+        }
+      case _ =>
+        None
+    }
+  }
+
+  /** Returns true if this path is a class directory. */
+  def isClassdir(path: String): Boolean =
+    dirs.contains(path)
+
+}
+
+object ClasspathIndex {
+  def empty: ClasspathIndex = ClasspathIndex(Classpath(Nil))
+  def apply(classpath: Classpath): ClasspathIndex = new Builder(classpath).result()
+
+  private final class Builder(classpath: Classpath) {
+    private val dirs = mutable.Map.empty[String, Classdir]
+
+    def result(): ClasspathIndex = {
+      val root = Classdir("/")
+      dirs(root.name) = root
+      classpath.entries.foreach(expandPath)
+      new ClasspathIndex(dirs)
+    }
+
+    private def expandPath(path: AbsolutePath): Unit = {
+      if (path.isFile) expandJar(path)
+      else if (path.isDirectory) expandDirectory(path)
+      else throw new IllegalArgumentException(path.toString)
+    }
+
+    private def getPackage(name: String): Classdir = {
+      dirs.get(name) match {
+        case Some(dir) =>
+          dir
+        case _ =>
+          val parent = getPackage(PathIO.dirname(name))
+          val entry = Classdir(name)
+          parent.members(PathIO.basename(name)) = entry
+          dirs(name) = entry
+          entry
+      }
+    }
+
+    private def expandJar(jarpath: AbsolutePath): Unit = {
```

Comment:
How about calling this method `expandJarEntry` and the method below `expandDirEntry`? That will disambiguate `directory` in `expandDirectory` between a directory entry and a directory element.

Reply:
Good idea, done.